Import libraries and dataset

In [2]:
import pandas as pd

df = pd.read_csv('simplified_coffee.csv')
df


,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,Ethiopia Shakiso Mormora,Revel Coffee,Medium-Light,United States,Ethiopia,4.70,92,November 2017,"Crisply sweet, cocoa-toned. Lemon blossom, roa..."
1,Ethiopia Suke Quto,Roast House,Medium-Light,United States,Ethiopia,4.19,92,November 2017,"Delicate, sweetly spice-toned. Pink peppercorn..."
2,Ethiopia Gedeb Halo Beriti,Big Creek Coffee Roasters,Medium,United States,Ethiopia,4.85,94,November 2017,"Deeply sweet, subtly pungent. Honey, pear, tan..."
3,Ethiopia Kayon Mountain,Red Rooster Coffee Roaster,Light,United States,Ethiopia,5.14,93,November 2017,"Delicate, richly and sweetly tart. Dried hibis..."
4,Ethiopia Gelgelu Natural Organic,Willoughby's Coffee & Tea,Medium-Light,United States,Ethiopia,3.97,93,November 2017,"High-toned, floral. Dried apricot, magnolia, a..."
...,...,...,...,...,...,...,...,...,...
1241,Finca Patzibir,El Gran Cafe,Medium-Light,Guatemala,Guatemala,4.70,92,November 2022,"Crisply sweet, nut-toned. Almond brittle, pie ..."
1242,Proyecto Aurora,El Gran Cafe,Medium-Light,Guatemala,Guatemala,2.94,93,November 2022,"Chocolaty, floral-framed. Cocoa nib, honeysuck..."
1243,Finca El Potrero,El Gran Cafe,Medium-Light,Guatemala,Guatemala,2.94,93,November 2022,"High-toned, enticingly sweet. Black cherry, na..."
1244,Chacayá Santiago Atitlán,El Gran Cafe,Medium-Light,Guatemala,Guatemala,4.70,93,November 2022,"Vibrantly sweet, subtly nuanced. Apricot, dar..."


Get an overview of the data in the next few cells.

In [2]:
df.head()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,Ethiopia Shakiso Mormora,Revel Coffee,Medium-Light,United States,Ethiopia,4.70,92,November 2017,"Crisply sweet, cocoa-toned. Lemon blossom, roa..."
1,Ethiopia Suke Quto,Roast House,Medium-Light,United States,Ethiopia,4.19,92,November 2017,"Delicate, sweetly spice-toned. Pink peppercorn..."
2,Ethiopia Gedeb Halo Beriti,Big Creek Coffee Roasters,Medium,United States,Ethiopia,4.85,94,November 2017,"Deeply sweet, subtly pungent. Honey, pear, tan..."
3,Ethiopia Kayon Mountain,Red Rooster Coffee Roaster,Light,United States,Ethiopia,5.14,93,November 2017,"Delicate, richly and sweetly tart. Dried hibis..."
4,Ethiopia Gelgelu Natural Organic,Willoughby's Coffee & Tea,Medium-Light,United States,Ethiopia,3.97,93,November 2017,"High-toned, floral. Dried apricot, magnolia, a..."


Explore the columns

In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1246 entries, 0 to 1245
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         1246 non-null   str    
 1   roaster      1246 non-null   str    
 2   roast        1234 non-null   str    
 3   loc_country  1246 non-null   str    
 4   origin       1246 non-null   str    
 5   100g_USD     1246 non-null   float64
 6   rating       1246 non-null   int64  
 7   review_date  1246 non-null   str    
 8   review       1246 non-null   str    
dtypes: float64(1), int64(1), str(7)
memory usage: 87.7 KB


In [3]:
df.shape

(1246, 9)

There might be 12 null values in the roast column.

Now, I need to check data types, which I can see above, but I will write code for it anyway.

In [4]:
print(df.dtypes)

name               str
roaster            str
roast              str
loc_country        str
origin             str
100g_USD       float64
rating           int64
review_date        str
review             str
dtype: object


There is a good mix of strings and numbers (floats and integers).  I might need to separate the individual words out of the review column for analysis.    

Next, I will check for null values.  

In [5]:
df.isnull()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...
1241,False,False,False,False,False,False,False,False,False
1242,False,False,False,False,False,False,False,False,False
1243,False,False,False,False,False,False,False,False,False
1244,False,False,False,False,False,False,False,False,False


Since the df only shows me the first and last 5 records, I don't know which rows contain the null values.

In [6]:
df.isna()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...
1241,False,False,False,False,False,False,False,False,False
1242,False,False,False,False,False,False,False,False,False
1243,False,False,False,False,False,False,False,False,False
1244,False,False,False,False,False,False,False,False,False


df.isna() and df.isnull() do the same thing and don't tell me what I need to know.   

Next, I will try this:

In [ ]:
null_rows = df[df['roast'].isnull()]
print(null_rows)

                                                   name  \
132                      Esmeralda Estate Panama Geisha   
480                      Esmeralda Estate Panama Geisha   
481               Asfaw Maru Ethiopia Natural Cold Brew   
482                      Ethiopia Nano Challa Cold Brew   
483        Brazil Ipanema Black Edition A-41 Red Cherry   
606                      Esmeralda Estate Panama Geisha   
822            Brazil Ipanema Golden Edition C26 Lychee   
826                      Esmeralda Estate Panama Geisha   
882   Ethiopia Guji Hambela Wate 74110/74112/74158 M...   
886                                       Yemen Lot 106   
892       Ethiopia Guji Shakisso Mormora Estate Natural   
1131                     Esmeralda Estate Panama Geisha   

                     roaster roast    loc_country    origin  100g_USD  rating  \
132        Difference Coffee   NaN        England    Panama    111.11      94   
480        Difference Coffee   NaN        England    Panama    111.11 

There are twelve null rows.  I am going to drop them since it's just a few rows and I have enough data otherwise.  I could fill in the missing data with the most common value (type of roast) but it might be inaccurate.  To drop the rows:

In [10]:
df.dropna()

,name,roaster,roast,loc_country,origin,100g_USD,rating,review_date,review
0,Ethiopia Shakiso Mormora,Revel Coffee,Medium-Light,United States,Ethiopia,4.70,92,November 2017,"Crisply sweet, cocoa-toned. Lemon blossom, roa..."
1,Ethiopia Suke Quto,Roast House,Medium-Light,United States,Ethiopia,4.19,92,November 2017,"Delicate, sweetly spice-toned. Pink peppercorn..."
2,Ethiopia Gedeb Halo Beriti,Big Creek Coffee Roasters,Medium,United States,Ethiopia,4.85,94,November 2017,"Deeply sweet, subtly pungent. Honey, pear, tan..."
3,Ethiopia Kayon Mountain,Red Rooster Coffee Roaster,Light,United States,Ethiopia,5.14,93,November 2017,"Delicate, richly and sweetly tart. Dried hibis..."
4,Ethiopia Gelgelu Natural Organic,Willoughby's Coffee & Tea,Medium-Light,United States,Ethiopia,3.97,93,November 2017,"High-toned, floral. Dried apricot, magnolia, a..."
...,...,...,...,...,...,...,...,...,...
1241,Finca Patzibir,El Gran Cafe,Medium-Light,Guatemala,Guatemala,4.70,92,November 2022,"Crisply sweet, nut-toned. Almond brittle, pie ..."
1242,Proyecto Aurora,El Gran Cafe,Medium-Light,Guatemala,Guatemala,2.94,93,November 2022,"Chocolaty, floral-framed. Cocoa nib, honeysuck..."
1243,Finca El Potrero,El Gran Cafe,Medium-Light,Guatemala,Guatemala,2.94,93,November 2022,"High-toned, enticingly sweet. Black cherry, na..."
1244,Chacayá Santiago Atitlán,El Gran Cafe,Medium-Light,Guatemala,Guatemala,4.70,93,November 2022,"Vibrantly sweet, subtly nuanced. Apricot, dar..."


We now have only 1234 rows instead of the original 1246, so the df.dropna() was successful.  Let's do df.shape to see if it stays that way:

In [12]:
df.shape

(1246, 9)

It went back to the original number of rows: 1246.  I think I need to use inplace = True.

In [14]:
df.dropna(inplace = True)
df.shape

(1234, 9)

It worked.  Now you can see that there are only 1234 rows.  However, when comparing this analysis to the chocolate bar analysis, I discovered that I can assign the no_null dataframe to a variable so I don't modify the original dataframe.  Since I've already done inplace = True, I've already modified the original df, and I can't undo it. I will leave it as is.   

Now, I will check again for any null values in the entire dataframe.

In [3]:
df.isnull().sum()

name            0
roaster         0
roast          12
loc_country     0
origin          0
100g_USD        0
rating          0
review_date     0
review          0
dtype: int64

It is still showing 12 null values in the 'roast' column.